clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -1040,84 +1124,156 @@ protected void produceToStoreBufferServiceOrKafka(
-           * {@link StoreVersionState} is properly primed, as other functions below this point, but prior to being
-           * enqueued into the {@link StoreBufferService} rely on this state to be there.
-           */
-          processStartOfPush(
-              record.getValue(),
-              controlMessage,
-              record.getTopicPartition().getPartitionNumber(),
-              partitionConsumptionStateMap.get(topicPartition.getPartitionNumber()));
-        }
-      }
-
       // Check schema id availability before putting consumer record to drainer queue
       waitReadyToProcessRecord(record);
-      // This function may modify the original record in KME and it is unsafe to use the payload from KME directly after
-      // this call.
-      DelegateConsumerRecordResult delegateConsumerRecordResult = delegateConsumerRecord(
-          record,
-          topicPartition.getPartitionNumber(),
+
+      totalBytesRead += handleSingleMessage(
+          new PubSubMessageProcessedResultWrapper<>(record),
+          topicPartition,
+          partitionConsumptionState,
           kafkaUrl,
           kafkaClusterId,
           beforeProcessingPerRecordTimestampNs,
-          beforeProcessingBatchRecordsTimestampMs);
-      switch (delegateConsumerRecordResult) {
-        case QUEUED_TO_DRAINER:
-          long queuePutStartTimeInNS = metricsEnabled ? System.nanoTime() : 0;
-
-          // blocking call
-          storeBufferService.putConsumerRecord(
-              record,
-              this,
-              null,
-              topicPartition.getPartitionNumber(),
-              kafkaUrl,
-              beforeProcessingPerRecordTimestampNs);
+          beforeProcessingBatchRecordsTimestampMs,
+          metricsEnabled,
+          elapsedTimeForPuttingIntoQueue);
+    }
 
-          if (metricsEnabled) {
-            elapsedTimeForPuttingIntoQueue += LatencyUtils.getElapsedTimeFromNSToMS(queuePutStartTimeInNS);
-          }
-          break;
-        case PRODUCED_TO_KAFKA:
-        case SKIPPED_MESSAGE:
-        case DUPLICATE_MESSAGE:
-          break;
-        default:
-          throw new VeniceException(
-              ingestionTaskName + " received unknown DelegateConsumerRecordResult enum for "
-                  + record.getTopicPartition());
+    /**
+     * Even if the records list is empty, we still need to check quota to potentially resume partition
+     */
+    storageUtilizationManager.enforcePartitionQuota(topicPartition.getPartitionNumber(), totalBytesRead);
+
+    if (metricsEnabled) {
+      if (totalBytesRead > 0) {
+        hostLevelIngestionStats.recordTotalBytesReadFromKafkaAsUncompressedSize(totalBytesRead);
+      }
+      if (elapsedTimeForPuttingIntoQueue.getValue() > 0) {
+        hostLevelIngestionStats.recordConsumerRecordsQueuePutLatency(
+            elapsedTimeForPuttingIntoQueue.getValue(),
+            beforeProcessingBatchRecordsTimestampMs);
       }
```

Comment:
Why carry this all the way here? Can we do the metrics recording directly inside `handleSingleMessage`, rather than merely setting the value into this `ValueHolder` over there?

Reply:
This is trying to preserve the existing semantics to report the latency per batch..
I don't mind to record metrics directly in the new function, but it would cause a change in the monitoring.

Reply:
Ah, so this metric is per batch or per record?

Reply:
It is per batch today:
[elapsedTimeForPuttingIntoQueue](https://github.com/linkedin/venice/blob/main/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java#L1106)

Reply:
Ok I see... this seems like a confusing metric semantic though xD